File: src/accountStateTimeline.ts
/**
 * Account-state TIMELINE (TRDD-YQZ9P8IL) — attribute every request to the subscription state active
 * AT THAT TIME, without hammering the SSD.
 *
 * The subscription state is a SLOWLY-CHANGING dimension: mode/plan almost never change; the account
 * changes on rotation (~minutes); only the 5h/7d % move continuously. So the write-frugal model is an
 * append-only STATE TIMELINE keyed on the DISCRETE dims, written ONLY when they change — a few
 * records/hour, not per-request. The continuously-moving 5h/7d % are DELIBERATELY excluded from the
 * change-detection key (they'd make the timeline write constantly); query them live instead.
 *
 * "Mode at request time T" = binary-search the timeline for the last record with ts <= T. Exact, and
 * needs ZERO per-request write.
 *
 * SSD discipline (the same doctrine as the span store): an in-memory buffer coalesces the rare bursts;
 * a flush appends + fsyncs the WHOLE batch once (NEVER per record — per-record fsync is the SSD
 * killer). Flush triggers, whichever first: a 60-second timer (tunable via AGENTLENS_ACCOUNT_STATE_FLUSH_MS),
 * the buffer reaching 32 records / ~16 KB, or graceful shutdown (the server pipes SIGTERM here). A
 * kill-9 mid-window loses at most the last window's changes and never a flushed record; NDJSON is
 * recovery-robust (a torn final line is simply discarded on read).
 */

import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import { classifyTtlRegime, type TtlContext } from './shared/cacheTtl'
import type { AccountInfo } from './accountInfo'

/** One state-timeline record — the subscription state that became active at `ts`. The discrete dims
 *  (everything except ts) are what change-detection keys on; the 5h/7d % are intentionally absent. */
export interface AccountStateRecord {
  ts: number                 // ms epoch when this state was first observed (the change moment)
  accountId: string | null
  email: string | null
  mode: string               // human billing mode (subscription within plan / usage credits / API)
  plan: string               // "Max 5x" | "Max 20x" | "Pro" | …
  authRegime: string         // raw regime: subscription | usage-credits | api-key | unknown
  ttlMinutes: number         // the cache-TTL tier of the main session under this state
  ttlSource: string          // doc-matrix | config | measured | assumed
}

/** TRDD-YQZ9P8IL — the human plan name from planType + the rate-limit tier's multiplier. Shared with
 *  get_account_status (mcpServer imports it) so the plan string has ONE source of truth. */
export function describePlan(planType: string | null | undefined, rateLimitTier: string | null | undefined): string {
  const mult = (rateLimitTier ?? '').toLowerCase().match(/(\d+)x\b/)  // default_claude_max_5x → "5"
  const suffix = mult ? ` ${mult[1]}x` : ''
  switch ((planType ?? '').toLowerCase()) {
    case 'max': return `Max${suffix}`
    case 'pro': return 'Pro'
    case 'team': return 'Team'
    case 'enterprise': return 'Enterprise'
    case 'free': return 'Free'
    default: return planType ? `${planType}${suffix}` : (mult ? `Max ${mult[1]}x` : 'unknown')
  }
}

/** TRDD-YQZ9P8IL — the billing MODE in human words from the resolved auth regime. Shared with
 *  get_account_status. The substring-'subscription' fallback is the stripe_subscription fix. */
export function describeAccountMode(authRegime: string | null): string {
  switch (authRegime) {
    case 'subscription': return 'subscription (within plan)'
    case 'usage-credits': return 'subscription drawing usage credits (over plan limit)'
    case 'api-key': return 'API key (pay-per-token)'
    default: return 'unresolved'
  }
}

/** Resolve the raw auth regime from ttlCtx (preferred) or a coarse billingType read (substring match
 *  so stripe_subscription is not misread as api-key). null billingType → null (never a guess). */
export function resolveAuthRegimeLabel(account: AccountInfo | null, ttlCtx: TtlContext | null): string | null {
  return ttlCtx?.auth
    ?? (account?.billingType
      ? (account.billingType.toLowerCase().includes('subscription') ? 'subscription' : 'api-key')
      : null)
}

/** Build the current state record from the live account + TTL context. `now` is injected (never
 *  Date.now() inside) so the sampler and tests control the timestamp. */
export function buildAccountStateRecord(account: AccountInfo | null, ttlCtx: TtlContext | null, now: number): AccountStateRecord {
  const authRegime = resolveAuthRegimeLabel(account, ttlCtx)
  const ttlRegime = classifyTtlRegime('main', ttlCtx)  // get_account_status is a main-conversation tool
  return {
    ts: now,
    accountId: account?.accountUuid ?? null,
    email: account?.email ?? account?.label ?? null,
    mode: describeAccountMode(authRegime),
    plan: account && account.source !== 'none' ? describePlan(account.planType, account.rateLimitTier) : 'unknown',
    authRegime: authRegime ?? 'unknown',
    ttlMinutes: ttlRegime.ttlAssumedMin,
    ttlSource: ttlRegime.ttlSource,
  }
}

/** The discrete change-detection key — everything EXCEPT the continuously-moving parts (ts, and the
 *  5h/7d % which are never in the record). A  join can't collide with any field's own content. */
function discreteKey(r: AccountStateRecord): string {
  return [r.accountId ?? '∅', r.mode, r.plan, r.authRegime, r.ttlMinutes].join('')
}

/** Resolves the timeline file path. AGENTLENS_ACCOUNT_STATE_LOG overrides; else ~/.agentlens/account-state.ndjson. */
export function accountStateTimelinePath(): string {
  return process.env['AGENTLENS_ACCOUNT_STATE_LOG'] || path.join(os.homedir(), '.agentlens', 'account-state.ndjson')
}

/**
 * The change-detected, buffered, fsync-on-flush timeline writer. One instance per server process.
 * record() is called often (the burn tick, ~every 4s) but only ENQUEUES on a discrete-state change,
 * so the common case is a cheap key comparison with no allocation and no write.
 */
export class AccountStateTimeline {
  private readonly filePath: string
  private readonly flushMs: number
  private buffer: AccountStateRecord[] = []
  private bufferedBytes = 0
  private lastKey: string | null    // last ENQUEUED discrete key (seeded from the file tail on start)
  private flushTimer: ReturnType<typeof setInterval> | null = null
  private static readonly FLUSH_MAX_RECORDS = 32
  private static readonly FLUSH_MAX_BYTES = 16 * 1024

  constructor(opts?: { filePath?: string; flushMs?: number; autoTimer?: boolean }) {
    this.filePath = opts?.filePath ?? accountStateTimelinePath()
    this.flushMs = opts?.flushMs ?? Math.max(1000, Number(process.env['AGENTLENS_ACCOUNT_STATE_FLUSH_MS']) || 60_000)
    // Seed lastKey from the file tail so a server RESTART into an unchanged state does not re-log it.
    this.lastKey = this.readLastKey()
    if (opts?.autoTimer !== false) {
      this.flushTimer = setInterval(() => this.flush(), this.flushMs)
      this.flushTimer.unref?.()  // never keep the process alive just for this timer
    }
  }

  /** Enqueue the record IFF its discrete state differs from the last enqueued one. Returns true when a
   *  change was recorded, false for the (common) no-change case — the change-detection that makes this
   *  a few-writes/hour timeline instead of a per-request firehose. */
  record(state: AccountStateRecord): boolean {
    const key = discreteKey(state)
    if (key === this.lastKey) return false
    this.lastKey = key
    this.buffer.push(state)
    this.bufferedBytes += JSON.stringify(state).length + 1
    if (this.buffer.length >= AccountStateTimeline.FLUSH_MAX_RECORDS || this.bufferedBytes >= AccountStateTimeline.FLUSH_MAX_BYTES) {
      this.flush()
    }
    return true
  }

  /** Append the buffered batch as NDJSON + fsync ONCE. Best-effort: a write failure re-buffers the
   *  batch (bounded — state changes are rare) rather than losing it, and never throws. No-op when empty. */
  flush(): void {
    if (this.buffer.length === 0) return
    const batch = this.buffer
    this.buffer = []
    this.bufferedBytes = 0
    const lines = batch.map(r => JSON.stringify(r)).join('\n') + '\n'
    try {
      fs.mkdirSync(path.dirname(this.filePath), { recursive: true })
      const fd = fs.openSync(this.filePath, 'a')          // O_APPEND: atomic append at EOF
      try {
        fs.writeSync(fd, lines)
        fs.fsyncSync(fd)                                   // durability once per BATCH, never per record
      } finally {
        fs.closeSync(fd)
      }
    } catch {
      // Transient write failure — put the batch back in front of anything enqueued meanwhile so the
      // next flush retries it in order. Bounded in practice because discrete changes are rare.
      this.buffer = batch.concat(this.buffer)
      this.bufferedBytes += lines.length
    }
  }

  /** Stop the timer and do a final flush — the graceful-shutdown path (piggybacks the server's SIGTERM). */
  stop(): void {
    if (this.flushTimer) {
      clearInterval(this.flushTimer)
      this.flushTimer = null
    }
    this.flush()
  }

  private readLastKey(): string | null {
    try {
      const raw = fs.readFileSync(this.filePath, 'utf8')
      const lines = raw.split('\n').filter(l => l.trim())
      if (lines.length === 0) return null
      return discreteKey(JSON.parse(lines[lines.length - 1]) as AccountStateRecord)
    } catch {
      return null   // no file yet, or an unreadable/torn tail — start fresh, never throw
    }
  }
}

/**
 * Resolve the account state active at time T: the last record with ts <= T (binary search over the
 * append-ordered timeline). Reads the whole NDJSON — it is tiny (a few records/hour). A torn final
 * line is skipped. null when no record precedes T (the timeline doesn't reach that far back).
 */
export function resolveStateAt(ts: number, filePath = accountStateTimelinePath()): AccountStateRecord | null {
  let records: AccountStateRecord[]
  try {
    const raw = fs.readFileSync(filePath, 'utf8')
    records = raw.split('\n')
      .filter(l => l.trim())
      .map(l => { try { return JSON.parse(l) as AccountStateRecord } catch { return null } })
      .filter((r): r is AccountStateRecord => r !== null && typeof r.ts === 'number')
  } catch {
    return null
  }
  let lo = 0, hi = records.length - 1, ans = -1
  while (lo <= hi) {
    const mid = (lo + hi) >> 1
    if (records[mid].ts <= ts) { ans = mid; lo = mid + 1 } else { hi = mid - 1 }
  }
  return ans >= 0 ? records[ans] : null
}

File: rust-core/crates/agentlens-core/src/account_state_timeline.rs
//! Port of the shared helpers in src/accountStateTimeline.ts (TRDD-YQZ9P8IL, ported under
//! TRDD-DMWOBWFH P4x.2c) — the human PLAN name, the human billing MODE, and the raw auth regime.
//!
//! They live in their own module for the same reason they do in TS: `get_account_status` and the
//! account-state sampler must produce the SAME plan and mode strings, and the way to guarantee that
//! is one implementation, not two that agree today.
//!
//! `resolve_state_at` and `build_account_state_record` are both ported below; only
//! `get_account_state_at` (the tool wrapper around them) is not.

use std::io::Write as _;
use std::path::PathBuf;
use std::sync::OnceLock;

use regex::Regex;
use serde_json::{Map, Value};

use crate::burn::account_info::AccountInfo;
use crate::burn::cache_ttl::{classify_ttl_regime, SessionTtlKind, TtlContext};
use crate::summarize::helpers;

/// `default_claude_max_5x` → "5". Hoisted: compiling a Regex inside a per-call function is the
/// clippy `regex_creation_in_loops` shape and pointless work on a hot path.
fn tier_multiplier_re() -> &'static Regex {
    static RE: OnceLock<Regex> = OnceLock::new();
    RE.get_or_init(|| Regex::new(r"(\d+)x\b").unwrap())
}

/// The human plan name from planType + the rate-limit tier's multiplier.
///
/// The DEFAULT arm is not a catch-all "unknown": an unrecognised `planType` is echoed VERBATIM with
/// the multiplier, because a plan name this code has not heard of is still the truth about the
/// account, and replacing it with "unknown" would hide a real answer behind a placeholder.
pub fn describe_plan(plan_type: Option<&str>, rate_limit_tier: Option<&str>) -> String {
    let tier = rate_limit_tier.unwrap_or("").to_lowercase();
    let mult = tier_multiplier_re().captures(&tier).map(|c| c[1].to_owned());
    // ` 5x`, not ` 5` — the `x` belongs to the SUFFIX in the TS, so `Max` + suffix is "Max 5x".
    let suffix = mult.as_ref().map_or(String::new(), |m| format!(" {m}x"));
    match plan_type.unwrap_or("").to_lowercase().as_str() {
        "max" => format!("Max{suffix}"),
        "pro" => "Pro".to_owned(),
        "team" => "Team".to_owned(),
        "enterprise" => "Enterprise".to_owned(),
        "free" => "Free".to_owned(),
        _ => match plan_type.filter(|p| !p.is_empty()) {
            // TS truthiness: an EMPTY planType is falsy and falls to the multiplier branch.
            Some(p) => format!("{p}{suffix}"),
            None => mult.map_or_else(|| "unknown".to_owned(), |m| format!("Max {m}x")),
        },
    }
}

/// The billing MODE in human words from the resolved auth regime.
pub fn describe_account_mode(auth_regime: Option<&str>) -> &'static str {
    match auth_regime {
        Some("subscription") => "subscription (within plan)",
        Some("usage-credits") => "subscription drawing usage credits (over plan limit)",
        Some("api-key") => "API key (pay-per-token)",
        _ => "unresolved",
    }
}

/// The raw auth regime from ttlCtx (preferred) or a coarse billingType read.
///
/// The SUBSTRING match is load-bearing: the real value is `stripe_subscription`, and an equality
/// check against "subscription" misreads a paying subscriber as an api-key account — which flips
/// the whole cache-TTL model (1h tier vs 5m) and every cost number derived from it.
/// A null billingType yields null, never a guess.
pub fn resolve_auth_regime_label(account: Option<&AccountInfo>, ttl_ctx: Option<&TtlContext>) -> Option<String> {
    // `ttlCtx?.auth ?? …` — the TS TtlContext always carries an `auth`, so a present ctx always
    // wins, including its 'unknown'.
    if let Some(ctx) = ttl_ctx {
        return Some(ctx.auth.as_str().to_owned());
    }
    let billing = account?.billing_type.as_deref().filter(|b| !b.is_empty())?;
    Some(if billing.to_lowercase().contains("subscription") { "subscription" } else { "api-key" }.to_owned())
}

/// `AGENTLENS_ACCOUNT_STATE_LOG || dataPath('account-state.ndjson')` — a FALSY-or, so an empty
/// override falls through to the data dir rather than resolving to the empty path.
pub fn account_state_timeline_path(data_dir: &std::path::Path) -> std::path::PathBuf {
    match std::env::var("AGENTLENS_ACCOUNT_STATE_LOG") {
        Ok(v) if !v.is_empty() => std::path::PathBuf::from(v),
        _ => data_dir.join("account-state.ndjson"),
    }
}

/// The NDJSON state timeline, oldest first. A missing or unreadable file is an EMPTY timeline, never
/// an error: "we have never observed an account" is a real answer, and throwing here would turn a
/// first run into a failure. A torn line is skipped individually — one bad record must not discard
/// the history around it — and a record without a numeric `ts` is dropped because the binary search
/// below is only sound on an ordered, dated list.
pub fn read_timeline(path: &std::path::Path) -> Vec<serde_json::Value> {
    let Ok(text) = std::fs::read_to_string(path) else { return Vec::new() };
    text.split('\n')
        .filter(|l| !l.trim().is_empty())
        .filter_map(|l| serde_json::from_str::<serde_json::Value>(l).ok())
        .filter(|r| r.get("ts").and_then(serde_json::Value::as_f64).is_some())
        .collect()
}

/// The newest record at or BEFORE `ts` — the account state in force at that moment.
///
/// Binary search over the append-ordered timeline: `records[mid].ts <= ts` moves the answer forward,
/// so the result is the LAST record that does not postdate the query. `None` means the timeline does
/// not reach back that far, which the caller must report as a gap rather than as "no account".
pub fn resolve_state_at(ts: f64, path: &std::path::Path) -> Option<serde_json::Value> {
    let records = read_timeline(path);
    let t = |r: &serde_json::Value| r.get("ts").and_then(serde_json::Value::as_f64).unwrap_or(f64::NAN);
    let (mut lo, mut hi, mut ans) = (0i64, records.len() as i64 - 1, -1i64);
    while lo <= hi {
        let mid = (lo + hi) >> 1;
        if t(&records[mid as usize]) <= ts {
            ans = mid;
            lo = mid + 1;
        } else {
            hi = mid - 1;
        }
    }
    (ans >= 0).then(|| records[ans as usize].clone())
}

/// Build the current state record — key order `ts, accountId, email, mode, plan, authRegime,
/// ttlMinutes, ttlSource` matches the TS wire shape exactly (the parity fixtures compare by key,
/// but keeping order identical makes a raw diff readable too). `now_ms` is injected, never read
/// from a clock here, so callers (and tests) control the timestamp.
pub fn build_account_state_record(account: Option<&AccountInfo>, ttl_ctx: Option<&TtlContext>, now_ms: f64) -> Map<String, Value> {
    let auth_regime = resolve_auth_regime_label(account, ttl_ctx);
    // Always Main: get_account_status is a main-conversation tool, never subagent/fork — passing
    // the subagent kind here would silently halve every TTL reported.
    let ttl_regime = classify_ttl_regime(Some(SessionTtlKind::Main), ttl_ctx);
    let mut m = Map::new();
    m.insert("ts".into(), helpers::num(now_ms));
    m.insert("accountId".into(), account.and_then(|a| a.account_uuid.clone()).map_or(Value::Null, Value::from));
    // `email ?? label` — label is a plain String (always present on a real account), so a `None`
    // email falls to it even when label is empty; but a present EMPTY-STRING email is kept as-is
    // (`??`, not `||`) rather than replaced by the label.
    m.insert(
        "email".into(),
        account.map_or(Value::Null, |a| Value::from(a.email.clone().unwrap_or_else(|| a.label.clone()))),
    );
    m.insert("mode".into(), Value::from(describe_account_mode(auth_regime.as_deref())));
    // The planType is deliberately IGNORED when source == "none" — an account synthesized as a
    // placeholder has no real plan to report, however plausible its planType field looks.
    m.insert(
        "plan".into(),
        Value::from(match account {
            Some(a) if a.source != "none" => describe_plan(a.plan_type.as_deref(), a.rate_limit_tier.as_deref()),
            _ => "unknown".to_owned(),
        }),
    );
    m.insert("authRegime".into(), Value::from(auth_regime.unwrap_or_else(|| "unknown".to_owned())));
    m.insert("ttlMinutes".into(), helpers::num(ttl_regime.ttl_assumed_min));
    m.insert("ttlSource".into(), Value::from(ttl_regime.ttl_source));
    m
}

/// The discrete change-detection key. `email` and `ttlSource` are deliberately EXCLUDED — email
/// changes independently of the billing state, and including ttlSource would write a record every
/// time it flips assumed<->measured, turning a few-writes/hour timeline into a firehose. A missing
/// accountId collapses to the literal sentinel "∅" (a documented, intentional collision with any
/// record that literally carries that string). `\x01` can't appear in any of these fields, unlike
/// a comma.
fn discrete_key(r: &Map<String, Value>) -> String {
    let account_id = match r.get("accountId") {
        Some(Value::String(s)) => s.clone(),
        _ => "∅".to_owned(),
    };
    let s = |k: &str| r.get(k).and_then(Value::as_str).unwrap_or("").to_owned();
    let ttl_minutes = r.get("ttlMinutes").and_then(Value::as_f64).map_or_else(String::new, |n| helpers::num(n).to_string());
    [account_id, s("mode"), s("plan"), s("authRegime"), ttl_minutes].join("\u{1}")
}

/// The change-detected, buffered, fsync-on-flush timeline writer — no auto timer here (alcore's
/// own chore drives `flush()` on its schedule, same split as every other chore in this crate).
pub struct AccountStateTimeline {
    file_path: PathBuf,
    buffer: Vec<Map<String, Value>>,
    buffered_bytes: usize,
    last_key: Option<String>,
}

impl AccountStateTimeline {
    const FLUSH_MAX_RECORDS: usize = 32;
    const FLUSH_MAX_BYTES: usize = 16 * 1024;

    /// Seeds `last_key` from the LAST line of the file, so a restart into an unchanged state does
    /// not re-log it. Any failure (absent file, unreadable, torn tail) seeds `None` rather than
    /// erroring — a fresh timeline is a legitimate starting state, not a fault.
    pub fn open(file_path: PathBuf) -> Self {
        let last_key = std::fs::read_to_string(&file_path).ok().and_then(|text| {
            text.split('\n')
                .rfind(|l| !l.trim().is_empty())
                .and_then(|line| serde_json::from_str::<Map<String, Value>>(line).ok())
                .map(|r| discrete_key(&r))
        });
        Self { file_path, buffer: Vec::new(), buffered_bytes: 0, last_key }
    }

    /// Enqueue `state` iff its discrete key differs from the last enqueued one. Returns true when a
    /// change was recorded. Auto-flushes once either bound is hit.
    pub fn record(&mut self, state: Map<String, Value>) -> bool {
        let key = discrete_key(&state);
        if Some(&key) == self.last_key.as_ref() {
            return false;
        }
        self.last_key = Some(key);
        self.buffered_bytes += serde_json::to_string(&state).map_or(0, |s| s.len() + 1);
        self.buffer.push(state);
        if self.buffer.len() >= Self::FLUSH_MAX_RECORDS || self.buffered_bytes >= Self::FLUSH_MAX_BYTES {
            self.flush();
        }
        true
    }

    /// Append the buffered batch as NDJSON and fsync ONCE per batch — never per record, which is
    /// the SSD killer this design exists to avoid. A write failure re-buffers the batch IN FRONT of
    /// anything enqueued meanwhile (order preserved) and never panics — losing it would be silent
    /// data loss, and the buffer stays bounded in practice because discrete changes are rare.
    pub fn flush(&mut self) {
        if self.buffer.is_empty() {
            return;
        }
        let batch = std::mem::take(&mut self.buffer);
        self.buffered_bytes = 0;
        let mut lines = String::new();
        for r in &batch {
            if let Ok(s) = serde_json::to_string(r) {
                lines.push_str(&s);
                lines.push('\n');
            }
        }
        let result = (|| -> std::io::Result<()> {
            if let Some(dir) = self.file_path.parent() {
                std::fs::create_dir_all(dir)?;
            }
            let mut f = std::fs::OpenOptions::new().create(true).append(true).open(&self.file_path)?;
            f.write_all(lines.as_bytes())?;
            f.sync_all()
        })();
        if result.is_err() {
            self.buffered_bytes = lines.len();
            let mut restored = batch;
            restored.append(&mut self.buffer);
            self.buffer = restored;
        }
    }

    /// Records currently buffered (unflushed) — test seam.
    pub fn buffered(&self) -> usize {
        self.buffer.len()
    }
}
